File: falcon/tensor/tensor.cc
#include <iostream>

#include "tensor.h"
#include "falcon/functional/functional.h"

namespace falcon {

tensor::tensor(af::array&& data, const bool requires_grad) {
  tensor_data->data = data;      
  tensor_data->requires_grad = requires_grad;
  if (requires_grad)
    tensor_data->grad = af::constant(0, tensor_data->data.dims());
}

tensor::tensor(af::array&& data, const bool requires_grad, std::vector<tensor> parents, OpType op) {
  tensor_data->data = data;
  tensor_data->requires_grad = requires_grad;
  tensor_data->parents = parents;
  tensor_data->op = op;
  if (requires_grad)
    tensor_data->grad = af::constant(0, tensor_data->data.dims());
}

tensor::tensor(af::array&& data, const bool requires_grad, std::vector<tensor> parents, float mul, OpType op) {
  tensor_data->data = data;
  tensor_data->requires_grad = requires_grad;
  tensor_data->parents = parents;
  tensor_data->op = op;
  tensor_data->mul = mul;
  if (requires_grad)
    tensor_data->grad = af::constant(0, tensor_data->data.dims());
}

af::array& tensor::data() const {
  return tensor_data->data; 
}

af::array& tensor::grad() const {
  return tensor_data->grad; 
}

bool tensor::is_grad_on() const {
  if (tensor_data->requires_grad) 
    return true;
  return false;
}

tensor tensor::operator+(const tensor& other) {
  F f;
  return f.add(*this, other);
}

tensor tensor::operator-(const tensor& other) {
  F f;
  return f.sub(*this, other);
}

tensor tensor::operator/(const float number) {
  F f;
  return f.div(*this, number);
}

tensor tensor::operator*(const float number) {
  F f;
  return f.mul(*this, number);
}

tensor tensor::operator*(const tensor& other) {
  F f;
  return f.mul(*this, other);
}

tensor tensor::operator^(const float number) {
  F f;
  return f.pow(*this, number);
}

void tensor::add_backward(const af::array& output_grad) const {
  for (tensor& parent: tensor_data->parents) {
    if (parent.tensor_data->requires_grad)
      parent.tensor_data->grad += output_grad;
  }
}

void tensor::sub_backward(const af::array& output_grad) const {
  if (tensor_data->parents[0].tensor_data->requires_grad)
    tensor_data->parents[0].tensor_data->grad += output_grad;
  if (tensor_data->parents[1].tensor_data->requires_grad)
    tensor_data->parents[1].tensor_data->grad -= output_grad;
}

void tensor::mul_backward(const af::array& output_grad) const {
  if (tensor_data->parents.size() == 1)
    tensor_data->parents[0].tensor_data->grad += output_grad * tensor_data->mul;  
  else {
    std::vector<tensor> parents = tensor_data->parents;
    if (tensor_data->parents[0].tensor_data->requires_grad)
      tensor_data->parents[0].tensor_data->grad += parents[1].data() * output_grad;
    if (tensor_data->parents[1].tensor_data->requires_grad)
      tensor_data->parents[1].tensor_data->grad += parents[0].data() * output_grad;
  }
}

void tensor::div_backward(const af::array& output_grad) const {
  if (tensor_data->parents[0].tensor_data->requires_grad)
    tensor_data->parents[0].tensor_data->grad += output_grad / tensor_data->mul;
}

void tensor::pow_backward(const af::array& output_grad) const {
  if (tensor_data->parents[0].tensor_data->requires_grad) {
    tensor_data->parents[0].tensor_data->grad += output_grad * tensor_data->mul * \
      af::pow(tensor_data->parents[0].data(), tensor_data->mul - 1);
  }
}

void tensor::sin_backward(const af::array& output_grad) const {
  if (tensor_data->parents[0].tensor_data->requires_grad)
    tensor_data->parents[0].tensor_data->grad += \
      af::cos(tensor_data->parents[0].tensor_data->data) * output_grad; 
}

void tensor::cos_backward(const af::array& output_grad) const {
  if (tensor_data->parents[0].tensor_data->requires_grad)
    tensor_data->parents[0].tensor_data->grad -= \
      af::sin(tensor_data->parents[0].tensor_data->data) * output_grad; 
}

void tensor::backward(const tensor& curr_tensor, const af::array& output_grad) {
#define GRAD_OPTYPE() {                                               \
  switch (curr_tensor.tensor_data->op) {                              \
    case OpType::ADD: curr_tensor.add_backward(output_grad); break;   \
    case OpType::SUB: curr_tensor.sub_backward(output_grad); break;   \
    case OpType::MUL: curr_tensor.mul_backward(output_grad); break;   \
    case OpType::DIV: curr_tensor.div_backward(output_grad); break;   \
    case OpType::POW: curr_tensor.pow_backward(output_grad); break;   \
    case OpType::SIN: curr_tensor.sin_backward(output_grad); break;   \
    case OpType::COS: curr_tensor.cos_backward(output_grad); break;   \
  }                                                                   \
}
#define GRAD_FN() {                                                       \
  switch (curr_tensor.tensor_data->op) {                                  \
    case OpType::ADD: std::cout << "add_backward()" << std::endl; break;  \
    case OpType::SUB: std::cout << "sub_backward()" << std::endl; break;  \
    case OpType::MUL: std::cout << "mul_backward()" << std::endl; break;  \
    case OpType::DIV: std::cout << "div_backward()" << std::endl; break;  \
    case OpType::POW: std::cout << "pow_backward()" << std::endl; break;  \
    case OpType::SIN: std::cout << "sin_backward()" << std::endl; break;  \
    case OpType::COS: std::cout << "cos_backward()" << std::endl; break;  \
  }                                                                       \
}
  if (curr_tensor.tensor_data->visited || curr_tensor.tensor_data->parents.size() == 0) 
    return;
  curr_tensor.tensor_data->visited = true;
  GRAD_OPTYPE();
//#define DEBUGING_MODE
#ifdef DEBUGING_MODE
  GRAD_FN();
#endif
  for (tensor& parent: curr_tensor.tensor_data->parents)
    if (parent.tensor_data->requires_grad)
      backward(parent, parent.tensor_data->grad);
#undef GRAD_OPTYPE
#undef GRAD_FN
}

void tensor::backward(bool retain_graph) {
  // Implicit gradient creation
  tensor_data->grad = af::constant(1, data().dims());
  backward(*this, tensor_data->grad);
}

} // namespace falcon

File: volgar/tensor.py
import math

import numpy as np

from .common import *
from .ops import OpType 

class Tensor:
  def __init__(self, data, requires_grad=False, parents=None, mul=None, op=None):
    self.data = np.array(data)
    self.requires_grad = requires_grad
    self.grad = np.zeros_like(self.data)
    self.op = op
    self.parents = parents
    self.mul = mul
    self.visited = False

  def __repr__(self):
    return f"Tensor: {self.data}, requires_grad={self.requires_grad}"

  def size(self):
    return self.data.shape 

  def __add__(self, other):
    other = other if isinstance(other, Tensor) else Tensor(other)
    if other.size() == self.size():  
      return Tensor(self.data + other.data, self.requires_grad or other.requires_grad, parents=[self, other], op=OpType.ADD)
    error("Dimensions should be equal")

  def __sub__(self, other):
    other = other if isinstance(other, Tensor) else Tensor(other)
    if other.size() == self.size():  
      return Tensor(self.data - other.data, self.requires_grad or other.requires_grad, parents=[self, other], op=OpType.SUB)
    error("Dimensions should be equal")

  def __mul__(self, other):
    other = other if isinstance(other, Tensor) else Tensor(other)
    """
    if self.size()[-1] == other.size()[0]:
      # elementwise multiplication
      print("passed")
      return Tensor(np.dot(self.data, other.data), self.requires_grad or other.requires_grad, parents=[self, other], op=OpType.DOT)
    """
    return Tensor(self.data * other.data, self.requires_grad or other.requires_grad, parents=[self], mul=other.data, op=OpType.MUL)

  def __pow__(self, other):
    # elementwise power 
    # other: can be a one dimensional tensor or a natural number 
    other = other if isinstance(other, Tensor) else Tensor(other)
    return Tensor(self.data ** other.data, self.requires_grad or other.requires_grad, parents=[self], mul=other.data, op=OpType.POW)

  def backward(self):
    self.grad = np.ones_like(self.data)
    self._backward(self)

  def _backward(self, curr_tensor):
    if (curr_tensor.visited or curr_tensor.parents == None): return
    curr_tensor.visited = True
    #self._debug_grad_fn(curr_tensor)
    # taking bacward gradients
    self._grad_optype(curr_tensor)
    
    for parent in curr_tensor.parents:
      if parent.requires_grad: 
        self._backward(parent)

  def _grad_optype(self, tensor):
    match tensor.op:
      case OpType.ADD: self._add_backward(tensor)
      case OpType.SUB: self._sub_backward(tensor)
      case OpType.MUL: self._mul_backward(tensor)
      case OpType.DOT: self._dot_backward(tensor)
      case OpType.DIV: self._div_backward(tensor)
      case OpType.POW: self._pow_backward(tensor)
      case OpType.RELU: self._relu_backward(tensor)

  def _debug_grad_fn(self, tensor):
    # for debuging purposes
    match tensor.op:
      case OpType.ADD: print("add_backward") 
      case OpType.SUB: print("sub_backward") 
      case OpType.MUL: print("mul_backward") 
      case OpType.DIV: print("div_backward") 
      case OpType.POW: print("pow_backward") 

  # -- backward functions -- #
  # todo: add backward functions for activation functions

  def _add_backward(self, tensorj):
    for parent in tensor.parents:
      if parent.requires_grad: parent.grad += tensor.grad

  def _sub_backward(self, tensor):
    if tensor.parents[0].requires_grad: tensor.parents[0].grad += tensor.grad
    if tensor.parents[1].requires_grad: tensor.parents[1].grad -= tensor.grad

  def _mul_backward(self, tensor):
    if tensor.mul is not None:
      if tensor.parents[0].requires_grad: tensor.parents[0].grad += tensor.grad * tensor.mul
    """
    else:
      if self.parents[0].requires_grad: self.parents[0].grad += self.parents[1].data * grad
      if self.parents[1].requires_grad: self.parents[1].grad += self.parents[0].data * grad
    """

  # https://en.wikipedia.org/wiki/Vector_calculus_identities#cite_note-4:~:text=.-,Dot%20product%20rule,-%5Bedit%5D
  def _dot_backward(self, other):
    pass
    """
    if tensor.parents[0].requires_grad and tensor.parents[1].requires_grad:
      tensor.parents
    """

  def _div_backward(self, grad):
    if tensor.mul is not None:
      if tensor.parents[0].requires_grad: tensor.parents[0].grad += tensor.grad / tensor.mul
    """
    else:
      if self.parents[0].requires_grad: self.parents[0].grad += (1. / self.parents[1].data) * grad
      if self.parents[1].requires_grad: self.parents[1].grad += (self.parents[0].data / self.parents[1].data ** 2) * grad
    """

  def _pow_backward(self, tensor):
    # for operations where tensor to the power scalar
    if tensor.mul is not None:
      tensor.parents[0].grad += tensor.grad * tensor.mul * (tensor.parents[0].data ** (tensor.mul - 1))

  def _relu_backward(self, tensor):
    if tensor.parents[0].requires_grad:
      tensor.parents[0].grad += 0 if (tensor.data <= 0).all() == True else tensor.grad

  @classmethod
  def zeros(cls, *shape, **kwargs):
    return cls(np.zeros(shape, dtype=np.float32), **kwargs)

  # todo: look into kaiming initialization methods
  @classmethod
  def glorot_uniform(cls, *shape, **kwargs):
    return cls((np.random.default_rng().random(size=shape, dtype=np.float32) * 2 - 1) * ((6 / (shape[0] + math.prod(shape[1:]))) ** 0.5), **kwargs) 